src/restic/backend/gs/gs.go

```diff
@@ -0,0 +1,461 @@
+package gs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"restic"
+	"strings"
+	"time"
+
+	"restic/backend"
+	"restic/errors"
+	"restic/debug"
+
+	"io/ioutil"
+
+	"golang.org/x/oauth2/google"
+	storage "google.golang.org/api/storage/v1"
+	"google.golang.org/api/googleapi"
+)
+
+// Backend stores data on an gs endpoint.
+type Backend struct {
+	service    *storage.Service
+	projectID  string
+	location   string
+	sem        *backend.Semaphore
+	bucketName string
+	prefix     string
+	backend.Layout
+}
+
+// make sure that *Backend implements backend.Backend
+var _ restic.Backend = &Backend{}
+
+const defaultLayout = "default"
+
+func getStorageService(jsonKeyPath string) (*storage.Service, error) {
+
+	raw, err := ioutil.ReadFile(jsonKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	conf, err := google.JWTConfigFromJSON(raw, storage.DevstorageReadWriteScope)
+	if err != nil {
+		return nil, err
+	}
+
+	client := conf.Client(context.TODO())
+
+	service, err := storage.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	return service, nil
+}
+
+// Open opens the gs backend at bucket and region. The bucket is created if it
+// does not exist yet.
+func Open(cfg Config) (restic.Backend, error) {
+	debug.Log("open, config %#v", cfg)
+
+	service, err := getStorageService(cfg.JSONKeyPath)
+	if err != nil {
+		fmt.Println(70)
+		return nil, errors.Wrap(err, "getStorageService")
+	}
+
+	sem, err := backend.NewSemaphore(cfg.Connections)
+	if err != nil {
+		fmt.Println(76)
+		return nil, err
+	}
+
+	be := &Backend{
+		service:    service,
+		projectID:  cfg.ProjectID,
+		location:   cfg.Location,
+		sem:        sem,
+		bucketName: cfg.Bucket,
+		prefix:     cfg.Prefix,
+		Layout: &backend.DefaultLayout{
+			Path: cfg.Prefix,
+			Join: path.Join,
+		},
+	}
+
+	// Create bucket if not exists
+	if _, err := service.Buckets.Get(be.bucketName).Do(); err != nil {
+		bucket := &storage.Bucket{
+			Name: be.bucketName,
+			Location: be.location,
+		}
+		if _, err := service.Buckets.Insert(be.projectID, bucket).Do(); err != nil {
+			return nil, errors.Wrap(err, "service.Buckets.Insert")
+		}
+	}
+
+	return be, nil
+}
+
+// IsNotExist returns true if the error is caused by a not existing file.
+func (be *Backend) IsNotExist(err error) bool {
+	debug.Log("IsNotExist(%T, %#v)", err, err)
+	return os.IsNotExist(err)
```

Comment:
Is that enough? For the s3 backend, we've had to test for more things: https://github.com/restic/restic/blob/master/src/restic/backend/s3/s3.go#L103